src/target/riscv/riscv.c

```diff
@@ -655,20 +664,50 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
 		return ERROR_FAIL;
 	if (riscv_reg_get(target, &tdata2_rb, GDB_REGNO_TDATA2) != ERROR_OK)
 		return ERROR_FAIL;
-	bool tdata1_config_denied = (tdata1 & ~tdata1_ignore_mask) != (tdata1_rb & ~tdata1_ignore_mask);
-	bool tdata2_config_denied = tdata2 != tdata2_rb;
-	if (tdata1_config_denied || tdata2_config_denied) {
+
+	const uint32_t type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+	const bool is_mcontrol = type == CSR_TDATA1_TYPE_MCONTROL;
+
+	/* Determine if tdata1 supports what we need.
+	 * For mcontrol triggers, we don't care about
+	 * the value in the read-only "maskmax" field.
+	 */
+	const riscv_reg_t tdata1_ignore_mask = is_mcontrol ? CSR_MCONTROL_MASKMAX(riscv_xlen(target)) : 0;
+	const bool tdata1_config_denied = (tdata1 & ~tdata1_ignore_mask) != (tdata1_rb & ~tdata1_ignore_mask);
+
+	/* Determine if tdata1.maxmask is sufficient
+	 * (only relevant for mcontrol triggers and NAPOT match type)
+	 */
+	bool unsupported_napot_range = false;
+	const bool is_napot_match = get_field(tdata1_rb, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
+	riscv_reg_t maskmax_mask, maskmax_value;
+	if (is_mcontrol && is_napot_match) {
+		maskmax_mask = CSR_MCONTROL_MASKMAX(riscv_xlen(target));
+		maskmax_value = get_field(tdata1_rb, maskmax_mask);
```

Comment:
For safety, please initialize `maskmax_value`.

Variable `maskmax_mask` can be const. Alternatively, since it is now used in a single place only, it can be removed completely. (Please pick which solution you prefer.)

```suggestion
	riscv_reg_t maskmax_value = 0;
	if (is_mcontrol && is_napot_match) {
		const riscv_reg_t maskmax_mask = CSR_MCONTROL_MASKMAX(riscv_xlen(target));
		maskmax_value = get_field(tdata1_rb, maskmax_mask);
```
or

```suggestion
	riscv_reg_t maskmax_value = 0;
	if (is_mcontrol && is_napot_match) {
		maskmax_value = get_field(tdata1_rb, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
```

Reply:
addressed